File: 00_ETC/10866.cpp
#include <iostream>
#include <deque>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n;
    cin >> n;
    deque<int> q;

    for (int i = 0; i < n; ++i) {
        int x;
        string s;
        cin >> s;
        if (s == "push_back") {
            cin >> x;
            q.push_back(x);
        } else if (s == "push_front") {
            cin >> x;
            q.push_front(x);
        } else if (s == "pop_front") {
            if (q.empty()) {
                cout << -1 << endl;
            } else {
                cout << q.front() << endl;
                q.pop_front();
            }
        } else if (s == "pop_back") {
            if (q.empty()) {
                cout << -1 << endl;
            } else {
                cout << q.back() << endl;
                q.pop_back();
            }
        } else if (s == "front") {
            if (q.empty()) {
                cout << -1 << endl;
            } else {
                cout << q.front() << endl;
            }
        } else if (s == "back") {
            if (q.empty()) {
                cout << -1 << endl;
            } else {
                cout << q.back() << endl;
            }
        } else if (s == "empty") {
            cout << (q.empty() ? 1 : 0) << endl;
        } else if (s == "size") {
            cout << (int) q.size() << endl;
        }
    }
    return 0;
}
//
// Created by 최승우 on 25. 8. 12.
//

File: 00_ETC_PY/10866.py
from collections import deque
import sys

input = sys.stdin.readline

n = int(input())
q = deque()
out = []

for _ in range(n):
    cmd = input().strip().split()

    if cmd[0] == "push_back":
        q.append(int(cmd[1]))
    elif cmd[0] == "push_front":
        q.appendleft(int(cmd[1]))
    elif cmd[0] == "pop_front":
        out.append(str(q.popleft() if q else -1))
    elif cmd[0] == "pop_back":
        out.append(str(q.pop() if q else -1))
    elif cmd[0] == "front":
        out.append(str(q[0] if q else -1))
    elif cmd[0] == "back":
        out.append(str(q[-1] if q else -1))
    elif cmd[0] == "empty":
        out.append('1' if not q else '0')
    elif cmd[0] == "size":
        out.append(str(len(q)))

sys.stdout.write("\n".join(out))